src/viam/sdk/tests/test_resource.cpp

```diff
@@ -0,0 +1,172 @@
+#include <viam/sdk/resource/resource_api.hpp>
+#include <viam/sdk/spatialmath/geometry.hpp>
+#include <viam/sdk/spatialmath/orientation.hpp>
+#include <viam/sdk/spatialmath/orientation_types.hpp>
+
+namespace viam {
+namespace sdktests {
+
+using namespace viam::sdk;
+
+BOOST_AUTO_TEST_CASE(test_api) {
+    API api1;
+    api1.set_namespace("ns");
+    BOOST_CHECK_EQUAL(api1.type_namespace(), "ns");
+    api1.set_resource_type("component");
+    BOOST_CHECK_EQUAL(api1.resource_type(), "component");
+    api1.set_resource_subtype("st");
+    BOOST_CHECK_EQUAL(api1.resource_subtype(), "st");
+    BOOST_CHECK_EQUAL(api1.to_string(), "ns:component:st");
+    BOOST_CHECK(!api1.is_service_type());
+    BOOST_CHECK(api1.is_component_type());
+
+    API api2 = API::from_string("ns:service:st");
+    BOOST_CHECK_EQUAL(api2.to_string(), "ns:service:st");
+    BOOST_CHECK(api2.is_service_type());
+    BOOST_CHECK(!api2.is_component_type());
+
+    BOOST_CHECK_THROW(API::from_string("nsservicest"), std::string);
+    BOOST_CHECK_THROW(API::from_string("ns:service:#st"), std::string);
+    BOOST_CHECK_THROW(API::from_string("ns:service"), std::string);
+}
+
+BOOST_AUTO_TEST_CASE(test_name) {
+    Name name1(API::from_string("ns:service:st"), "remote", "name");
+    BOOST_CHECK_EQUAL(name1.api().to_string(), "ns:service:st");
+    BOOST_CHECK_EQUAL(name1.remote_name(), "remote");
+    BOOST_CHECK_EQUAL(name1.name(), "name");
+    BOOST_CHECK_EQUAL(name1.to_string(), "ns:service:st/remote:name");
+    BOOST_CHECK_EQUAL(name1.short_name(), "remote:name");
+    // TODO: Currently, remotes not being returned from proto for whatever reason
+    // BOOST_CHECK_EQUAL(Name::from_proto(name1.to_proto()), name1);
+
+    Name name2 = Name::from_string("ns:component:st/name");
+    BOOST_CHECK_EQUAL(name2.api().to_string(), "ns:component:st");
+    BOOST_CHECK_EQUAL(name2.short_name(), "name");
+    BOOST_CHECK_EQUAL(Name::from_proto(name2.to_proto()), name2);
+
+    BOOST_CHECK_THROW(Name::from_string("ns:service:#st/remote:name"), std::string);
+}
+
+BOOST_AUTO_TEST_CASE(test_model) {
+    ModelFamily mf("ns", "mf");
+    BOOST_CHECK_EQUAL(mf.to_string(), "ns:mf");
+
+    Model model1(mf, "model1");
+    BOOST_CHECK_EQUAL(model1.to_string(), "ns:mf:model1");
+    Model model2("ns", "mf", "model2");
+    BOOST_CHECK_EQUAL(model2.to_string(), "ns:mf:model2");
+
+    Model model3 = Model::from_str("ns:mf:model3");
+    BOOST_CHECK_EQUAL(model3.to_string(), "ns:mf:model3");
+    Model model4 = Model::from_str("model4");
+    BOOST_CHECK_EQUAL(model4.to_string(), "rdk:builtin:model4");
+
+    ModelFamily empty("", "");
+    Model model5(empty, "model5");
+    BOOST_CHECK_EQUAL(model5.to_string(), "model5");
+
+    BOOST_CHECK_THROW(Model::from_str("@"), std::runtime_error);
+}
+
+BOOST_AUTO_TEST_CASE(test_linkconfig) {
+    translation t;
+    t.x = 0;
+    t.y = 1;
+    t.z = 2;
+    OrientationConfig ocfg = OrientationConfig();
+    GeometryConfig gcfg;
```

Comment:
I suspect (but have not tested) that the errors causing test failure are centered around this value. Looking through the logs we see the following error message: `/tmp/src/viam/sdk/../../viam/sdk/spatialmath/geometry.hpp:42:7: runtime error: load of value 3983782352, which is not a valid value for type 'GeometryType'`.

What this reads as to me is the `GeometryType` value we're getting is random junk data rather than a default value. This makes some sense, because `GeometryType` is just an enum so an int under the hood. From some brief glancing around, `GeometryType` is really only used in `GeometryConfig`, and the other places where we're looking at `GeometryConfig` we're either getting it from a proto type (in which case we'd set the `GeometryType` in conversion) or we're explicitly setting it. 

I suspect if we set some sort of default value for `geometry_type_` when constructing a `GeometryConfig` then the errors (or at least the ones I saw) would go away. This is definitely something we'll want to see fixed before merging this PR. If you want to test locally and not go through the very slow turnaround process of committing and checking CI, you can use the `dockerfile`s in the SDK to imitate the CI process locally. Let me know if you want any help with that, happy to provide it!

Reply:
`GeometryConfig` has no constructors at all, so all trivial fields like `geometry_type_` will be uninitialized when a `GeometryConfig` is declared on the stack. I suggest giving it a default constructor (or non-static data member initializers!) for such fields. Worth checking other related types like `pose` for similar issues.

Generally though, I dislike having a "default" for things like this. Notably `OrientationConfig` defaults to a quaternion with a particular quaternion-y value. But why that default and not some other?

My suggestion would be to 1) paper over the issue by adding some defaults to `GeometryConfig`, and then filing a follow-up ticket to take a closer look at the initialization/defaults/API of the spatial math and related types, ideally with a goal of avoiding the need for arbitrary defaults.

Reply:
Done! Made a [ticket](https://viam.atlassian.net/browse/RSDK-6379).